build-logic/jvm/src/main/kotlin/gradlebuild.unittest-and-compile.gradle.kts

```diff
@@ -139,13 +139,27 @@ fun addDependencies() {
 
 fun addCompileAllTask() {
     tasks.register("compileAll") {
+        description = "Compile all source code, including main, test, integTest, crossVersionTest, testFixtures, etc."
         val compileTasks = project.tasks.matching {
             it is JavaCompile || it is GroovyCompile
         }
         dependsOn(compileTasks)
     }
+
+    tasks.register("compileAllProduction") {
+        description = "Compile all production source code, usually only main and testFixtures."
+        val compileTasks = project.tasks.matching {
+            (it is JavaCompile || it is GroovyCompile) && !it.isTestCompile()
+        }
+        dependsOn(compileTasks)
+    }
 }
 
+val testCompileTasks
+    get() = TestType.values().map { "compile${it.prefix.toUpperCase()}Test" } + listOf("compileTestGroovy", "compileTestKotlin")
```

Comment:
There is also the `jmh` compilation, I wonder if it makes sense to exclude that as well or if it is quick enough. Since we don't run those tests anywhere, maybe it makes sense to exclude them.

Reply:
Should we remove the JMH code altogether maybe? We shouldn't leave the code in there if it's not at least compiled as some part of the CI pipeline.

Reply:
We should compile it at some point in the CI pipeline. I think we probably should run the JMH tests once a week so we can look up the results if we want to.

For now, I suppose it would be fine to compile the jmh tests as part of sanity check.

Reply:
I'm not sure about "once a week". What if that build is broken, who has incentive to fix it? I think we should make it a requirement that we can only release Gradle if all the code compiles.

Reply:
This was a misunderstanding:
- compile jmh tests -> on every commit
- run jmh tests -> once a week

Reply:
It seems fine compiling `jmh` because it takes at most 1s: https://ge.gradle.org/s/lbji7uch6b2n6/timeline?sort=longest&task-path=jmh